sstables/trie.cc

```diff
@@ -0,0 +1,787 @@
+#include <immintrin.h>
+#include "sstables/index_reader.hh"
+#include "utils/small_vector.hh"
+#include "trie_writer.hh"
+#include "trie_serializer.hh"
+
+namespace trie {
+
+using namespace trie;
+
+// A pointer (i.e. file position) to a node on disk, with some initial metadata parsed. 
+struct reader_node {
+    struct page_ptr : cached_file::ptr_type {
+        using parent = cached_file::ptr_type;
+        page_ptr() noexcept = default;
+        page_ptr(parent&& x) noexcept : parent(std::move(x)) {}
+        page_ptr(const page_ptr& other) noexcept : parent(other ? other->share() : nullptr) {}
+        page_ptr(page_ptr&&) noexcept = default;
+        page_ptr& operator=(page_ptr&&) noexcept = default;
+        page_ptr& operator=(const page_ptr& other) noexcept {
+            parent::operator=(other ? other->share() : nullptr);
+            return *this;
+        }
+    };
+    // Position of this node in the input file.
+    size_t pos;
+    // Number of children preemptively extracted from the representation.
+    uint16_t n_children;
+    // Payload bits preemptively extracted from the representation.
+    uint8_t payload_bits;
+
+    // Looks up the first child of this node greater with transition greater or equal to the given one.
+    // If such child doesn't exist, the `idx` of the result will be negative.
+    lookup_result lookup(std::byte transition, const_bytes page);
+    // Looks up the child with the given index.
+    // If there is no child with such an index (can happen in DENSE nodes, which have empty slots),
+    // picks the closest child with idx greater (if `forward == true`) or smaller (if `forward == false`)
+    // than the given. If there is no such child, the `idx` of the result will be negative.
+    lookup_result get_child(int idx, bool forward, const_bytes page);
+    // Returns a view of the payload of this node.
+    // The `bytes` view can extend beyond 
+    payload_result payload(const_bytes page) const;
+    // Returns a view of the raw on-disk representation of the node.
+    const_bytes raw(const_bytes page) const;
+};
+
+// For iterating over the trie, we need to keep a stack of those,
+// forming a path from the root to the current node.
+//
+// child_idx is the index of the child currently visited by the containing trie_cursor.
+// child_idx might be equal to -1, this means that the node itself, not any of its children,
+// is currently visited. 
+struct node_cursor {
+    reader_node node;
+    int child_idx;
+};
+
+template <typename T>
+concept trie_reader_source = requires(T& o, uint64_t pos, reader_permit rp) {
+    { o.read_page(pos) } -> std::same_as<future<std::pair<cached_file::ptr_type, bool>>>;
+    { o.read_row_index_header(pos, rp) } -> std::same_as<future<row_index_header>>;
+    { o.read_row_index_header(pos, rp) } -> std::same_as<future<row_index_header>>;
+};
+
+// Parses some basic type-oblivious metadata.
+static reader_node pv_to_reader_node(size_t pos, const_bytes page) {
```

Comment:
What is "pv"?

Reply:
cached_file::page_view. The name is outdated.